mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/model.scala

```diff
@@ -0,0 +1,1175 @@
+            val returnCode = ControlPacketFlags(v.getByte & 0xff)
+            decodeReturnCodes(remainingLen - (packetLenAtTopicFilter - v.len), returnCodes :+ returnCode)
+          } else {
+            returnCodes
+          }
+        val returnCodes = decodeReturnCodes(l - (packetLen - v.len), Vector.empty)
+        val returnCodesValid = returnCodes.nonEmpty && returnCodes.foldLeft(true) {
+          case (true, rc) if rc.underlying < ControlPacketFlags.QoSReserved.underlying => true
+          case _ => false
+        }
+        if (returnCodesValid) {
+          Right(SubAck(packetId, returnCodes))
+        } else {
+          Left(BadSubAckMessage(packetId, returnCodes))
+        }
+      } catch {
+        case _: NoSuchElementException => Left(BufferUnderflow)
+      }
+
+    // 3.10 UNSUBSCRIBE – Unsubscribe from topics
+    def decodeUnsubscribe(l: Int): Either[DecodeError, Unsubscribe] =
+      try {
+        val packetLen = v.len
+        val packetId = PacketId(v.getShort & 0xffff)
+        @tailrec
+        def decodeTopicFilters(
+            remainingLen: Int,
+            topicFilters: Vector[Either[DecodeError, String]]
+        ): Vector[Either[DecodeError, String]] =
+          if (remainingLen > 0) {
+            val packetLenAtTopicFilter = v.len
+            val topicFilter = v.decodeString()
+            decodeTopicFilters(remainingLen - (packetLenAtTopicFilter - v.len), topicFilters :+ topicFilter)
+          } else {
+            topicFilters
+          }
+        val topicFilters = decodeTopicFilters(l - (packetLen - v.len), Vector.empty)
+        val topicFiltersValid = topicFilters.nonEmpty && topicFilters.foldLeft(true) {
+          case (true, Right(_)) => true
+          case _ => false
+        }
+        if (topicFiltersValid) {
+          Right(Unsubscribe(packetId, topicFilters.flatMap {
+            case Right(tfs) => List(tfs)
+            case _ => List.empty
+          }))
+        } else {
+          Left(BadUnsubscribeMessage(packetId, topicFilters))
+        }
+      } catch {
+        case _: NoSuchElementException => Left(BufferUnderflow)
+      }
+
+    // 3.11 UNSUBACK – Unsubscribe acknowledgement
+    def decodeUnsubAck(): Either[DecodeError, UnsubAck] =
+      try {
+        val packetId = PacketId(v.getShort & 0xffff)
+        Right(UnsubAck(packetId))
+      } catch {
+        case _: NoSuchElementException => Left(BufferUnderflow)
+      }
+  }
+}
+
+object Command {
+
+  /**
+   * Send a command to an MQTT session
+   * @param command The command to send
+   */
+  def apply(command: ControlPacket): Command[Nothing] =
+    new Command(command)
+
+  /**
+   * Send a command to an MQTT session with data to carry through into
+   * any related event.
+   * @param command The command to send
+   * @param carry The data to carry through
+   * @tparam A The type of data to carry through
+   */
+  def apply[A](command: ControlPacket, carry: A): Command[A] =
+    new Command(command, carry)
+}
```

Comment:
Would it make sense to offer `create` methods for the important `ControlPacket` subtypes to get `Command.pubAck(packetId)`?

Reply:
Again, I'm conflicted about `create`. The Java developer can use `new` here, quite nicely.

Reply:
This would work nice for Scala, as well: instead of `Command(Connect(clientId,...))` you'd have `Command.connect(clientId)`
But yeah, just an idea.